subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c

```diff
@@ -0,0 +1,546 @@
+#include <net/socket.h>
+#include <net/cloud.h>
+#include <net/cloud_backend.h>
+#include <at_cmd.h>
+#include <stdio.h>
+
+#if defined(CONFIG_AWS_FOTA)
+#include <net/aws_fota.h>
+#endif
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(bifravst_cloud, CONFIG_BIFRAVST_CLOUD_LOG_LEVEL);
+
+#if defined(CONFIG_BIFRAVST_CLOUD_IPV6)
+#define BIFRAVST_CLOUD_AF_FAMILY AF_INET6
+#else
+#define BIFRAVST_CLOUD_AF_FAMILY AF_INET
+#endif
+
+#define AWS_CLOUD_CLIENT_ID_LEN 15
+
+#define AWS "$aws/things/"
+#define AWS_LEN (sizeof(AWS) - 1)
+
+#define SHADOW_BASE_TOPIC AWS "%s/shadow"
+#define SHADOW_BASE_TOPIC_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 7)
+
+#define ACCEPTED_TOPIC AWS "%s/shadow/get/accepted"
+#define ACCEPTED_TOPIC_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 20)
+
+#define REJECTED_TOPIC AWS "%s/shadow/get/rejected"
+#define REJECTED_TOPIC_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 20)
+
+#define CFG_TOPIC AWS "%s/shadow/get/accepted/desired/cfg"
+#define CFG_TOPIC_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 32)
+
+#define UPDATE_DELTA_TOPIC AWS "%s/shadow/update/delta"
+#define UPDATE_DELTA_TOPIC_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 20)
+
+#define UPDATE_TOPIC AWS "%s/shadow/update"
+#define UPDATE_TOPIC_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 14)
+
+#define SHADOW_GET AWS "%s/shadow/get"
+#define SHADOW_GET_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 11)
+
+#define BATCH_TOPIC AWS "%s/batch"
+#define BATCH_TOPIC_LEN (AWS_LEN + AWS_CLOUD_CLIENT_ID_LEN + 6)
+
+#define SUBSCRIBE_ID 1234
+
+static char client_id_buf[AWS_CLOUD_CLIENT_ID_LEN + 1];
+static char shadow_base_topic[SHADOW_BASE_TOPIC_LEN + 1];
+static char accepted_topic[ACCEPTED_TOPIC_LEN + 1];
+static char cfg_topic[CFG_TOPIC_LEN + 1];
+static char rejected_topic[REJECTED_TOPIC_LEN + 1];
+static char update_delta_topic[UPDATE_DELTA_TOPIC_LEN + 1];
+static char update_topic[UPDATE_TOPIC_LEN + 1];
+static char get_topic[SHADOW_GET_LEN + 1];
+static char batch_topic[BATCH_TOPIC_LEN + 1];
+
+static const struct mqtt_topic cc_rx_list[] = {
+	{ .topic = {
+		.utf8 = rejected_topic, .size = REJECTED_TOPIC_LEN },
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE },
+	{ .topic = {
+		.utf8 = cfg_topic, .size = CFG_TOPIC_LEN },
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE },
+	{ .topic = {
+		.utf8 = update_delta_topic, .size = UPDATE_DELTA_TOPIC_LEN },
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE },
+};
+
+struct bifravst_cloud_tx_data {
+	char *buf;
+	size_t len;
+	char *topic;
+	enum mqtt_qos qos;
+};
+
+static char rx_buffer[CONFIG_BIFRAVST_CLOUD_MQTT_RX_TX_BUFFER_LEN];
+static char tx_buffer[CONFIG_BIFRAVST_CLOUD_MQTT_RX_TX_BUFFER_LEN];
+static char payload_buf[CONFIG_BIFRAVST_CLOUD_MQTT_PAYLOAD_BUFFER_LEN];
+
+static struct mqtt_client client;
+static struct sockaddr_storage broker;
+static struct cloud_backend *bifravst_cloud_backend;
+
+static int client_id_get(char *const id)
+{
+	int err;
+	char imei_buf[AWS_CLOUD_CLIENT_ID_LEN + 1];
+
+	err = at_cmd_write("AT+CGSN", imei_buf,
+		(AWS_CLOUD_CLIENT_ID_LEN + 5), NULL);
```

Comment:
Why is the buffer size provided to `at_cmd_write` larger than the actual `imei_buf` size? This can cause stack corruption.

The cleanest approach would be to provide `sizeof(imei_buf)` to the `at_cmd_write`. If you need a larger buffer, just make it larger.